department-of-veterans-affairs/caseflow
# Enable back-end sorting of tasks by appeal type

Resolves #11315 

Adds efficient caching of the attributes `is_aod` and `case_type` (where case type here means, original, post remand, etc) in the `cached_appeal_attributes` table for both Legacy and AMA Appeals. Adds the backend work for the sorting capability.

### Acceptance Criteria
- [x]  Back end correctly stores each appeal's aod status and case type
- [x]  Front end correctly shows the back end sorted tasks

### Testing Plan

1. Run the new migration with

   `bin/rails db:migrate`

1. Run the updated job in the rails console with

   ```
   > UpdateCachedAppealsAttributesJob.perform_now
   > CachedAppeal.find_by(appeal_id: 11, appeal_type: Appeal.name)
   [2019-08-13 15:52:56 -0400]   CachedAppeal Load (1.3ms)  SELECT  "cached_appeal_attributes".* FROM "cached_appeal_attributes" WHERE "cached_appeal_attributes"."appeal_id" = $1 AND "cached_appeal_attributes"."appeal_type" = $2 LIMIT $3  [["appeal_id", 11], ["appeal_type", "Appeal"], ["LIMIT", 1]]
   => #<CachedAppeal:0x0000561541c19740
    appeal_id: 11,
    appeal_type: "Appeal",
    docket_number: "190804-11",
    docket_type: "direct_review",
    vacols_id: nil,
    case_type: "original",
    is_aod: false,
    closest_regional_office_city: "St. Petersburg",
    veteran_name: "Hudson, Luciano",
    issue_count: 1>
   ```

1. Change user to BVALSPORER and go to the `/organizations/vlj-support` "On hold" tab
1. Toggle the case type sorting button
   1. Note the AOD cases stay at the top (ASC & DESC)
   1. Secondary sorting should follow Case Type (ASC & DESC)
   1. Tertiary sortiing should follow Docket Number (ASC & DESC)

*Schema Changes Only*
